commons/tenant-manager/s3/retained_storage.go

```
@@ -0,0 +1,604 @@
+	if validationErr != nil {
+		return ObjectMetadata{}, newRetainedStorageError("create-or-recover", validationErr, nil)
+	}
+
+	return validated, nil
+}
+
+// recoverRetainedVersion selects the sole exact retained version for resolvedKey.
+// The versions listing is prefix-based, so sibling keys such as resolvedKey+".bak"
+// can share pages with the exact key; listing paginates until every entry for the
+// exact key has been evaluated and stops once the sorted listing passes it.
+func (s *recoverableRetainedStorage) recoverRetainedVersion(ctx context.Context, resolvedKey string) (ObjectMetadata, error) {
```

golangci-lint (2.12.2)

[error] cognitive complexity 35 of func `(*recoverableRetainedStorage).recoverRetainedVersion` is high (> 30)

(gocognit)